gpflow/training/monitor/monitor.py

```diff
@@ -0,0 +1,790 @@
+        """
+        super().__init__(model, event_path)
+        all_summaries = additional_summaries or []
+        parameters = parameters or list(model.parameters)
+
+        # Add scalar parameters
+        all_summaries += [tf.summary.scalar(p.full_name, tf.reshape(p.constrained_tensor, []))
+                          for p in parameters if p.size == 1]
+
+        # Add non-scalar parameters
+        if not only_scalars:
+            all_summaries += [tf.summary.histogram(p.full_name, p.constrained_tensor)
+                              for p in parameters if p.size > 1]
+
+        # Add likelihood
+        all_summaries.append(tf.summary.scalar("optimisation/likelihood",
+                                               model._likelihood_tensor))
+
+        # Create the summary tensor
+        self._summary = tf.summary.merge(all_summaries)
+
+
+class LmlTensorBoardTask(BaseTensorBoardTask):
+    """
+    Monitoring task that creates a TensorBoard with just one scalar value -
+    the unbiased estimator of the evidence lower bound (ELBO or LML).
+
+    The LML is averaged over a number of minimatches. The input dataset is split into the specified
+    number of sequential minibatches such that every datapoint is used exactly once. The set of
+    minibatches doesn't change from one iteration to another.
+
+    The task can display the progress of calculating LML at each iteration (how many minibatches
+    are left to compute). For that the `tqdm' progress bar should be installed (pip install tqdm).
+    """
+
+    def __init__(self, model: Model, event_path: str, minibatch_size: Optional[int] = 100,
+                 display_progress: Optional[bool] = True) -> None:
+        """
+        :param model: Model tensor
+        :param event_path: Path of the event file where the summary protocol buffers will
+        be written to.
+        :param minibatch_size: Number of points per minibatch
+        :param display_progress: if True the task displays the progress of calculating LML.
+        """
+
+        super().__init__(model, event_path)
+        self._minibatch_size = minibatch_size
+        self._full_lml = tf.placeholder(settings.tf_float, shape=())
+        self._summary = tf.summary.scalar(model.name + '/full_lml', self._full_lml)
+        self._display_progress = display_progress
+
+    def run(self, context: MonitorContext, *args, **kwargs) -> None:
+
+        with params_as_tensors_for(self._model):
+            tf_x, tf_y = self._model.X, self._model.Y
+
+        if self._display_progress:  # pragma: no cover
+            try:
+                import tqdm
+                wrapper = tqdm.tqdm
+            except ImportError:
+                print("LML monitor task: to display progress install `tqdm`.")
+                wrapper = lambda x: x
+        else:   # pragma: no cover
+            wrapper = lambda x: x
+
+        lml = 0.0
+        num_batches = -(-len(self._model.X._value) // self._minibatch_size)  # round up
```

Comment:
Recommended: There is `math.ceil`.
```
num_values = len(self._model.X._value)
num_batches = math.ceil(num_values / self._minibatch_size)
```

Reply:
I guess // is more efficient - no conversion from int to float and back. Obviously it's just a question of code beauty.

Reply:
math.ceil returns a float so needs casting back to int but I think this is not going to be time-critical and in this case I would favour being explicit about what you mean.

Reply:
Changed.